flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserCalcitePlanner.java

```diff
@@ -1070,7 +1073,11 @@ private RelNode genFilterRelNode(
                             .convert(subQueryExpr)
                             .accept(funcConverter);
 
-            RelNode filterRel = LogicalFilter.create(srcRel, convertedFilterLHS);
+            RelNode filterRel =
```

Comment:
Please check whether this is enough for the fix.

```
            RelNode filterRel = LogicalFilter.create(srcRel, convertedFilterLHS);
            if (!RelOptUtil.getVariablesUsed(filterRel).isEmpty()) {
                filterRel =
                        LogicalFilter.create(
                                srcRel,
                                convertedFilterLHS,
                                ImmutableSet.copyOf(RelOptUtil.getVariablesSet(filterRel)));
            }
```

Reply:
Good point. But it's not enough for the fix.
When replace with this code, the exception will still be thrown for the following sql
`select * from src where src.key in (select c.key from (select * from src b where exists (select a.key from src a where b.value = a.value)) c);`